src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Resolvers/DynamicObjectResolver.cs

```diff
@@ -347,6 +347,11 @@ public static TypeInfo BuildType(DynamicAssembly assembly, Type type, bool force
                 return null;
             }
 
+            if (type.IsNotPublic)
+            {
+                throw new MessagePackSerializationException("Building dynamic formatter only allows public type. Type: " + type.FullName);
```

Comment:
Does `DynamicObjectResolverAllowPrivate` fail here too then? What are the limitations of this? Only non-public *members* but the type itself must be public? Are we sure this check doesn't cause failures in more cases than already failed before?

I know how to overcome this limitation. We can open an issue to track it if/when we need it. I just really thought non-public types already worked.

Reply:
AllowPrivate calls `BuildFormatterToDynamicMethod` so does not affect this.
Currently TypeBuilder is throwing an exception(on CreateTypeInfo), which is not what we intended.
AssemblyBuilder does not support access to NonPublic Type.
Therefore AllowPrivate uses DynamicMethod instead.

Finally I'll change to `if (!(type.IsPublic || type.IsNestedPublic))` because NonPublic does not detect nested private type.

Reply:
> AssemblyBuilder does not support access to NonPublic Type.

Well, *that* type does support accessing non-public types. The issue is when you're done building the assembly the CLR will reject it when loading it. But there's a way to fix that. We can actually make your dynamically generated code fully support non-public types. If you're interested in doing that file an issue to track it. I've done it on a few other projects already.

Reply:
Very interesting, it will brings to able to remove DynamicMethod code.
And then it might want to integrate with AllowPrivate.
However, it is breaking changes.

I think the behavior should not be changed at present.
So I think we should put verify code, but are you worried?

Reply:
Worried about this change of yours? A little, yes. Since as I recall, testing for public-ness isn't particularly straightforward. You already caught `IsNestedPublic` though, so _maybe_ that's it. I don't object to your change -- if it's a mistake we can always fix it when we catch it.

> However, it is breaking changes.

Are you talking about a breaking change that we wouldn't need a special `AllowPrivate` class any more so the regular class would support private members? Yes, that would be breaking. I don't mind that there are two classes. But they could share a *lot* more and yes avoid DynamicMethod. 

I filed #660 to track.